IR-2377 spin products

SamMazerIR opened the pull request:
## Summary
- Ignore drag and only use click in Interactables
- Helper functions for TransformComponent (right, up, forward)
- More comprehensive options for default interaction aliases (spatial, nonspatial, combined - clicks/triggers/E key)
- no longer setting drag to false on mouse up, should reflect state similarly to pressed

- camera orbit system now requires dragging to be true before it captures input and begins orbiting
- InteractableComponent now requires dragging to be false to capture (this allows orbiting if you started mouseDown over an interactable)
- EditorControlSystem no longer needs accumulator or checking mouse down vs up entity, just checks for no dragging on mouse up instead
- no longer starting GizmoPlane with visibleComponent, this is managed in the TransformGizmoControlComponent when a control entity is dragged
- simplified TransformGizmoControlComponent logical flow, now properly updates VisibleComponent on gizmoPlane

## References
paving the way for ecommerce ticket:
[https://tsu.atlassian.net/browse/IR-2377](https://tsu.atlassian.net/browse/IR-2377)

fixes for single-click selection (gizmo plane was interfering)
[https://tsu.atlassian.net/browse/IR-1293](https://tsu.atlassian.net/browse/IR-1293)

## QA Steps
Interactables should only react to a click, not a click & drag (there is a threshold for cursor travel before it is considered a drag)

SamMazerIR committed:
wip on spinning ecommerce, drag detection in interactables

SamMazerIR committed:
helper functions and API for transform up, right, fwd. Can still call as TransformComponent.up(entity) but we can also call it on an instance of a transformcomponent (e.g. transform.up() )

SamMazerIR committed:
Merge branch 'dev' into IR-2377-spin-products

HexaField commented on packages/spatial/src/transform/components/TransformComponent.ts:
```diff
@@ -227,10 +233,32 @@ export const TransformComponent = defineComponent({
   },
 
+  forward: (entity: Entity) => {
+    const matrixElements = getComponent(entity, TransformComponent).matrix.elements
```
we should have an optional vec3 passed in to these functions that the output is put onto, rather than a single hidden global

SamMazerIR replied:
changed

HexaField commented on packages/spatial/src/transform/components/TransformComponent.ts:
```diff
@@ -71,7 +74,10 @@ export const TransformComponent = defineComponent({
       matrix: new Matrix4(),
-      matrixWorld: new Matrix4()
+      matrixWorld: new Matrix4(),
+      forward: () => TransformComponent.forward(entity),
```
this is an anti pattern. component definitions should just store data, not any utility functions

SamMazerIR replied:
removed per our discussion

SamMazerIR committed:
removing the helper functions we're not allowed to write

SamMazerIR committed:
Merge branch 'IR-2377-spin-products' of github.com:EtherealEngine/etherealengine into IR-2377-spin-products

SamMazerIR committed:
updating how the helper functions work, removing the stored constants in favor of a returned "out" Vector

SamMazerIR committed:
Merge branch 'dev' into IR-2377-spin-products

the-reality-engineer commented on packages/engine/src/interaction/components/InteractableComponent.ts:
```diff
@@ -288,18 +295,54 @@ export const InteractableComponent = defineComponent({
+
+            //TODO store this threshold of 0.001 somewhere (drag threshold)
+            if (inputPointer && inputPointerPosition.distanceToSquared(inputPointer.position) < 0.001) {
+              console.log('distanceSquared = ' + inputPointerPosition.distanceToSquared(inputPointer.position))
```
Please remove the log here

the-reality-engineer commented on packages/engine/src/interaction/components/InteractableComponent.ts:
```diff
@@ -288,18 +295,54 @@ export const InteractableComponent = defineComponent({
+              console.log('distanceSquared = ' + inputPointerPosition.distanceToSquared(inputPointer.position))
+              callInteractCallbacks(entity)
+            } else {
+              console.log('not firing interact')
```
And here

the-reality-engineer commented on packages/spatial/src/input/components/InputComponent.ts:
```diff
@@ -47,7 +47,9 @@ import { InputSourceComponent } from './InputSourceComponent'
-  Interact: [MouseButton.PrimaryClick, XRStandardGamepadButton.Trigger, KeyboardButton.KeyE]
+  MergedInteract: [MouseButton.PrimaryClick, XRStandardGamepadButton.Trigger, KeyboardButton.KeyE],
+  SpatialInteract: [MouseButton.PrimaryClick, XRStandardGamepadButton.Trigger],
+  NonSpatialInteract: [KeyboardButton.KeyE]
```
Not a huge fan of this additional complexity; is this necessary?

SamMazerIR replied:
I'm open to updating this, I do want the ability to use a merged spatial interact though

the-reality-engineer replied:
I still find this confusing as an API surface. I think these aliases should represent categories of user intentions, not categories of input.

SamMazerIR committed:
wip on gizmo changes

SamMazerIR committed:
WIP on refactoring inputs

SamMazerIR committed:
Merge branch 'IR-2377-spin-products' of github.com:EtherealEngine/etherealengine into IR-2377-spin-products

SamMazerIR committed:
adding ordering to InputComponent's useExecuteWithInput subsystems as part of the new InputExecutionSystemGroup

SamMazerIR committed:
button states updates

SamMazerIR committed:
dragging and rotating states hammered out, need to test and replace the logic in InteractableComponent in favor of drag detection inside of productComponent instead

SamMazerIR committed:
massive simplification of dragging and rotating events per button, simplified interactablecomponent based on this change, enables ecommerce interactions, gizmo cleanup

SamMazerIR committed:
Merge branch 'dev' into IR-2377-spin-products

SamMazerIR committed:
Merge branch 'dev' into IR-2377-spin-products

SamMazerIR committed:
reverting interaction alias changes

SamMazerIR committed:
Merge branch 'dev' into IR-2377-spin-products

SamMazerIR committed:
Merge branch 'dev' into IR-2377-spin-products

SamMazerIR committed:
adding inputSource Entity to button state

SamMazerIR committed:
no longer setting drag to false on mouse up, should reflect state similarly to pressed
camera orbit system now requires dragging to be true before it captures input and begins orbiting
InteractableComponent now requires dragging to be false to capture (this allows orbiting if you started mouseDown over an interactable)
EditorControlSystem no longer needs accumulator or checking mouse down vs up entity, just checks for no dragging on mouse up instead
no longer starting GizmoPlane with visibleComponent, this is managed in the TransformGizmoControlComponent when a control entity is dragged
simplified TransformGizmoControlComponent logical flow, now properly updates VisibleComponent on gizmoPlane

SamMazerIR committed:
prevent garbage generation per frame when checking pointer position before dragging is true

SamMazerIR committed:
Merge branch 'dev' into IR-2377-spin-products
